main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
-        let i = ref b @ r;
-        let done = () -> deref i >= e;
-        let next = () -> {
-                let res = deref i;
-                if (res < e) {
-                    i := res + 1;
-                    res
-                }
-                else {
-                    bug!("Empty Iterator!")
-                }
+    pub def range(rc: Region[r], b: Int32, e: Int32): Iterator[Int32, ef, r] \ Write(r) =
+        let pos = ref b @ rc;
+        if (e <= b)
+            empty(rc)
+        else {
+            let f = () -> match (deref pos) {
+                case i if i < e => pos := i+1; Ans(i)
+                case _          => Done
             };
-        Iterator(done, next)
+            let iterF = () -> unsafe_cast f() as _ \ { ef, r };
+            Iterator(rc, iterF)
+        }
 
     ///
     /// Returns an iterator over an iterable with the element `x` repeated `n` times.
     ///
     /// Returns an empty iterator if `n < 0`.
     ///
-    @Lazy
-    pub def repeat(r: Region[r], n: Int32, x: a): Iterator[a, r] \ Write(r) =
-        let i = ref 0 @ r;
-        let done = () -> deref i >= n;
-        let next = () -> {
-              let j = deref i;
-              if (j < n) {
-                i := j + 1;
-                x
-              }
-              else {
-                bug!("Empty Iterator!")
-              }
-            };
-        Iterator(done, next)
+    pub def repeat(rc: Region[r], n: Int32, x: a): Iterator[a, ef, r] \ Write(r) =
+        let ix = ref n @ rc;
+        let f = () -> {
+            let i = deref ix;
+            if (i < 1)
+                Done
+            else {
+                ix := i-1;
+                Ans(x)
+            }
+        };
+        let iterF = () -> unsafe_cast f() as _ \ { ef, r };
+        Iterator(rc, iterF)
 
     ///
     /// Returns the sum of all elements in the iterator `iter`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def sum(iter: Iterator[Int32, r]): Int32 \ Read(r) =
+    pub def sum(iter: Iterator[Int32, ef, r]): Int32 \ { ef, Read(r) } =
         foldLeft((acc, x) -> acc + x, 0, iter)
 
     ///
     /// Returns the sum of all elements in the iterator `iter` according to the function `f`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def sumWith(f: a -> Int32 \ ef, iter: Iterator[a, r]): Int32 \ { ef, Read(r) } =
+    pub def sumWith(f: a -> Int32 \ ef1, iter: Iterator[a, ef, r]): Int32 \ { ef, ef1, Read(r) } =
         foldLeft((acc, x) -> acc + f(x), 0, iter)
 
     ///
     /// Returns the product of all elements in the iterator `iter`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def product(iter: Iterator[Int32, r]): Int32 \ Read(r) =
-        if (isEmpty(iter))
-            1
-        else
-            foldLeft((acc, x) -> acc * x, 1, iter)
+    pub def product(iter: Iterator[Int32, ef, r]): Int32 \ { ef, Read(r) } = match next(iter) {
```

Comment:
The old code also had a match and a fold, but I guess a fold would be enough?

Reply:
I think this is cleaned up now if it was one of the `sum` / `product` functions.

Reply:
I think my point was that the match seems redundant. We can just call foldLeft directly.